sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/grouping.scala

```diff
@@ -212,3 +212,29 @@ object GroupingID {
     if (SQLConf.get.integerGroupingIdEnabled) IntegerType else LongType
   }
 }
+
+
+object GroupingAnalytics {
+  def unapply(exprs: Seq[Expression]):
+  Option[(Seq[Seq[Expression]], Seq[Seq[Expression]], Seq[Expression])] = {
+
+    val (groupingSets, others) = exprs.partition(_.isInstanceOf[GroupingSet])
+    if (groupingSets.isEmpty) {
+      None
+    } else {
+      val groups =
+        groupingSets.flatMap(_.asInstanceOf[GroupingSet].groupByExprs) ++ others
```

Comment:
It looks we don't need the line break:
```
      val groups = groupingSets.flatMap(_.asInstanceOf[GroupingSet].groupByExprs) ++ others
```

Reply:
Since the cast `.asInstanceOf[GroupingSet]` appears three times, could you cast it only once at the beginning?